crates/ui/src/windows/archive_manager.rs

```diff
@@ -0,0 +1,560 @@
+                            cross_justify: true,
+                            ..Default::default()
+                        },
+                        |ui| {
+                            if save_promise.is_none()
+                                && ui
+                                    .add_enabled(
+                                        view.as_ref()
+                                            .is_some_and(|view| view.iter().next().is_some()),
+                                        egui::Button::new("Create from selected files"),
+                                    )
+                                    .clicked()
+                            {
+                                if let Some(view) = view {
+                                    let version = *version;
+                                    match Self::create_archive(view, version) {
+                                        Ok(file) => {
+                                            *save_promise =
+                                                Some(luminol_core::spawn_future(async move {
+                                                    file.save(
+                                                        match version {
+                                                            1 => "Game.rgssad",
+                                                            2 => "Game.rgss2a",
+                                                            3 => "Game.rgss3a",
+                                                            _ => unreachable!(),
+                                                        },
+                                                        "RGSSAD archives",
+                                                    )
+                                                    .await
+                                                }))
+                                        }
+                                        Err(e) => update_state.toasts.error(e.to_string()),
+                                    }
+                                }
+                            } else if save_promise.is_some() {
+                                ui.spinner();
+                            }
+                        },
+                    );
+                });
+
+                if let Some(p) = save_promise.take() {
+                    match p.try_take() {
+                        Ok(Ok(())) => {
+                            update_state.toasts.info("Created archive successfully!");
+                        }
+                        Ok(Err(e)) => {
+                            if !matches!(e, luminol_filesystem::Error::CancelledLoading) {
+                                update_state.toasts.error(e.to_string())
+                            }
+                        }
+                        Err(p) => *save_promise = Some(p),
+                    }
+                }
+            }
+        }
+    }
+
+    fn create_archive(
```

Comment:
Create archive and copy_files should be run on another thread (possibly using `Promise::spawn_blocking`?) to avoid freezing up the ui

Reply:
I implemented `AsyncRead + AsyncWrite + AsyncSeek` for host files by performing IO in another thread on native and asynchronously waiting on the oneshot channels on web, and the archive manager now creates and extracts archives asynchronously